File: app/users/forms/signup.py
from flask_wtf import FlaskForm
from wtforms import StringField, PasswordField, SubmitField
from wtforms.validators import DataRequired, Email, EqualTo, Regexp

def _is_email_valid(form, field):
    validate_email = Email(message='Enter a valid email.')
    if field.data:
        validate_email(form, field)

class SignupForm(FlaskForm):
    """User Sign-up Form."""
    username = StringField('User name', validators=[DataRequired()])
    email = StringField('Email', validators=[_is_email_valid])
    password = PasswordField('Password', validators=[DataRequired()])
    confirm = PasswordField('Confirm Your Password', validators=[
        DataRequired(), EqualTo('password', message='Passwords must match.')])
    phone = StringField('Phone', validators=[DataRequired()])
    atomy_id = StringField('Atomy ID', validators=[Regexp(r'(S|\d\d{7})?')])
    submit = SubmitField('Register')

File: app/users/static/js/signup.js
$(document).ready(function () {
    $("#signupForm").submit(function (event) {
        event.preventDefault();

        var formData = {
            username: $("#username").val(),
            password: $("#password").val(),
            confirm: $("#confirm").val(),
            email: $("#email").val(),
            phone: $("#phone").val(),
            atomy_id: $("#atomy_id").val()
        };
        $('.invalid-feedback').remove(); // Remove existing error messages
        $('.is-invalid').removeClass('is-invalid'); // Remove invalid class from all fields
        $('#signupError').hide(); // Hide the signup error message
        $.ajax({
            type: "POST",
            url: "/api/v1/user/signup", // Replace with your actual signup endpoint
            data: JSON.stringify(formData),
            contentType: "application/json; charset=utf-8",
            dataType: "json",
            success: function (data) {
                if (data.error) {
                    $("#signupError").text(data.error).show();
                } else {
                    window.location.href = '/';
                }
            },
            error: function (error) {
                var errorData = error.responseJSON;
                if (errorData.fieldErrors) {
                    errorData.fieldErrors.forEach(function (fieldError) {
                        // Display the error message next to the corresponding field
                        var field = $(`#${fieldError.name}`);
                        field.addClass("is-invalid");
                        field.after(`<div class="invalid-feedback">${fieldError.status}</div>`);
                    });
                }
                $("#signupError").text(`Sign up failed: ${errorData.error}`).show();
            }
        });
    });
});
